File: src/main/scala/jk_5/nailed/server/map/game/script/FileSystem.scala
package jk_5.nailed.server.map.game.script

import java.io.{BufferedReader, IOException, InputStream, InputStreamReader}
import java.util
import java.util.regex.Pattern

import jk_5.nailed.api.mappack.filesystem.IMount
import jk_5.nailed.api.util.Checks
import org.apache.commons.io.FilenameUtils

import scala.collection.mutable

/**
 * No description given
 *
 * @author jk-5
 */
object FileSystem {

  private def sanitizePath(path: String, allowWildcards: Boolean = false): String = {
    var p = path.replace('\\', '/')
    val specialChars = Array[Char]('"', ':', '<', '>', '?', '|')
    val cleanName = new StringBuilder
    for(i <- 0 until p.length){
      val c = path.charAt(i)
      if(c >= ' ' && util.Arrays.binarySearch(specialChars, c) < 0 && (allowWildcards || c != '*')){
        cleanName.append(c)
      }
    }

    p = cleanName.toString()

    val parts = p.split("/")
    val outputParts = new util.Stack[String]()
    for(i <- 0 until parts.length){
      val part = parts(i)
      if(part.length() != 0 && part != "."){
        if(part == ".."){
          if(!outputParts.empty()){
            val top = outputParts.peek()
            if(!"..".equals(top)){
              outputParts.pop()
            }else{
              outputParts.push("..")
            }
          }else{
            outputParts.push("..")
          }
        }else if(part.length() >= 255){
          outputParts.push(part.substring(0, 255))
        }else{
          outputParts.push(part)
        }
      }
    }

    val result = new mutable.StringBuilder("")
    val it = outputParts.iterator()
    while(it.hasNext){
      val part = it.next()
      result.append(part)
      if(it.hasNext){
        result.append('/')
      }
    }

    result.toString()
  }

  def contains(pathA: String, pathB: String): Boolean = {
    val a = sanitizePath(pathA)
    val b = sanitizePath(pathB)

    pathB != ".." && !pathB.startsWith("../") && (pathB == pathA || pathA.length == 0 || pathB.startsWith(pathA + "/"))
  }

  def toLocal(path: String, location: String): String = {
    val p = sanitizePath(path)
    val loc = sanitizePath(location)
    assert(contains(location, path))
    val local = p.substring(loc.length)
    if(local.startsWith("/")){
      local.substring(1)
    }else local
  }
}

class FileSystem {

  private val mounts = mutable.HashMap[String, MountWrapper]()
  private val openFiles = mutable.HashSet[IMountedFile]()

  def unload(){
    openFiles.foreach{ f => try{
      f.close()
    }catch{
      case e: IOException => //Nom nom nom
    }}
  }

  def unmount(path: String){
    val p = FileSystem.sanitizePath(path)
    if(mounts.contains(p)) this.mounts.remove(p)
  }

  def list(path: String): Array[String] = {
    val p = FileSystem.sanitizePath(path)
    val mount = getMount(p)
    val list = new util.ArrayList[String]()
    mount.list(p, list)
    for(m <- this.mounts.valuesIterator){
      if(getDirectory(m.location) == p){
        list.add(getName(m.location))
      }
    }
    val array = new Array[String](list.size())
    list.toArray(array)
    array
  }

  private def findIn(dir: String, matches: mutable.ArrayBuffer[String], pattern: Pattern){
    val list = this.list(dir)
    for(i <- 0 until list.length){
      val e = list(i)
      val epath = dir + "/" + e
      if(pattern.matcher(epath).matches()){
        matches += epath
      }
      if(isDir(epath)){
        findIn(epath, matches, pattern)
      }
    }
  }

  def find(wildPath: String): Array[String] = {
    val p = FileSystem.sanitizePath(wildPath)
    val pattern = Pattern.compile("^\\Q" + p.replaceAll("\\*", "\\\\E[^\\\\/]*\\\\Q") + "\\E$")
    val matches = mutable.ArrayBuffer[String]()
    findIn("", matches, pattern)
    matches.toArray
  }

  def getDirectory(path: String): String = {
    val p = FileSystem.sanitizePath(path)
    if(p.length == 0) return ".."
    val lastSlash = p.indexOf('/')
    if(lastSlash >= 0){
      p.substring(0, lastSlash)
    }else ""
  }

  def getName(path: String): String = {
    val p = FileSystem.sanitizePath(path)
    if(p.length == 0) return "root"
    val lastSlash = p.indexOf('/')
    if(lastSlash >= 0){
      FilenameUtils.removeExtension(p.substring(lastSlash + 1))
    }else FilenameUtils.removeExtension(p)
  }

  def getSize(path: String): Long = {
    val p = FileSystem.sanitizePath(path)
    val mount = getMount(p)
    mount.getSize(p)
  }

  def isDir(path: String): Boolean = {
    val p = FileSystem.sanitizePath(path)
    val mount = getMount(p)
    mount.isDirectory(p)
  }

  def exists(path: String): Boolean = {
    val p = FileSystem.sanitizePath(path)
    val mount = getMount(p)
    mount.exists(p)
  }

  def openAsInputStream(path: String): InputStream = {
    val p = FileSystem.sanitizePath(path)
    val mount = getMount(p)
    val stream = mount.openForRead(p)
    if(stream == null) return null
    val ret = new InputStream with IMountedFile {
      override def read() = stream.read()
      override def read(b: Array[Byte]) = stream.read(b)
      override def read(b: Array[Byte], off: Int, len: Int) = stream.read(b, off, len)
      override def skip(n: Long) = stream.skip(n)
      override def available() = stream.available()
      override def mark(readlimit: Int) = stream.mark(readlimit)
      override def reset() = stream.reset()
      override def markSupported() = stream.markSupported()
      override def close(){
        openFiles -= this
        stream.close()
      }
    }
    openFiles += ret
    ret
  }

  def openForTextRead(path: String): IMountedFileText = {
    val p = FileSystem.sanitizePath(path)
    val mount = getMount(p)
    val stream = mount.openForRead(p)
    if(stream == null) return null
    val reader = new BufferedReader(new InputStreamReader(stream))
    val file = new IMountedFileText {
      override def readLine() = reader.readLine()
      override def close(){
        openFiles -= this
        reader.close()
      }
    }
    openFiles += file
    file
  }

  def openForBinaryRead(path: String): IMountedFileBinary = {
    val p = FileSystem.sanitizePath(path)
    val mount = getMount(p)
    val stream = mount.openForRead(p)
    if(stream == null) return null
    val file = new IMountedFileBinary {
      override def read() = stream.read()
      override def close(){
        openFiles -= this
        stream.close()
      }
    }
    openFiles += file
    file
  }

  def mount(location: String, mount: IMount){
    Checks.notNull(location, "Location is null")
    Checks.notNull(mount, "Mount is null")
    val loc = FileSystem.sanitizePath(location)
    if(loc.contains("..")){
      throw new FileSystemException("Cannot mount below root")
    }
    this.mounts.put(loc, new MountWrapper(loc, mount))
  }

  private def getMount(path: String): MountWrapper = {
    var m: MountWrapper = null
    var mL: Int = 999
    for(mount <- this.mounts.valuesIterator){
      if(FileSystem.contains(mount.location, path)){
        val len = FileSystem.toLocal(path, mount.location).length
        if(m == null || len < mL){
          m = mount
          mL = len
        }
      }
    }
    if(m == null) throw new FileSystemException("Invalid path")
    m
  }
}

File: src/main/java/jk_5/nailed/server/map/game/script/FileSystem.java
package jk_5.nailed.server.map.game.script;

import com.google.common.collect.Lists;
import com.google.common.collect.Maps;
import com.google.common.collect.Sets;
import jk_5.nailed.api.mappack.filesystem.IMount;
import jk_5.nailed.api.util.Checks;
import org.apache.commons.io.FilenameUtils;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.util.*;
import java.util.regex.Pattern;

/**
 * No description given
 *
 * @author jk-5
 */
public class FileSystem {

    private Map<String, MountWrapper> mounts = Maps.newHashMap();
    private Set<IMountedFile> openFiles = Sets.newHashSet();

    public void unload() {
        //noinspection SynchronizeOnNonFinalField
        synchronized(this.openFiles){
            while(this.openFiles.size() > 0){
                IMountedFile file = this.openFiles.iterator().next();
                try{
                    file.close();
                }catch(IOException e){
                    this.openFiles.remove(file);
                }
            }
        }
    }

    public synchronized void mount(String location, IMount mount) throws FileSystemException {
        if(mount == null){
            throw new NullPointerException();
        }
        location = sanitizePath(location);
        if(location.contains("..")){
            throw new FileSystemException("Cannot mount below the root");
        }
        mount(new MountWrapper(location, mount));
    }

    private synchronized void mount(MountWrapper wrapper) throws FileSystemException {
        String location = wrapper.getLocation();
        if(this.mounts.containsKey(location)){
            this.mounts.remove(location);
        }
        this.mounts.put(location, wrapper);
    }

    @SuppressWarnings("unused")
    public synchronized void unmount(String path) {
        path = sanitizePath(path);
        if(this.mounts.containsKey(path)){
            this.mounts.remove(path);
        }
    }

    public synchronized String combine(String path, String childPath) {
        path = sanitizePath(path);
        childPath = sanitizePath(childPath);

        if(path.length() == 0){
            return childPath;
        }
        if(childPath.length() == 0){
            return path;
        }
        return sanitizePath(path + '/' + childPath);
    }

    public static String getDirectory(String path) {
        path = sanitizePath(path);
        if(path.length() == 0){
            return "..";
        }

        int lastSlash = path.lastIndexOf('/');
        if(lastSlash >= 0){
            return path.substring(0, lastSlash);
        }
        return "";
    }

    public static String getName(String path) {
        path = sanitizePath(path);
        if(path.length() == 0){
            return "root";
        }

        int lastSlash = path.lastIndexOf('/');
        if(lastSlash >= 0){
            return FilenameUtils.removeExtension(path.substring(lastSlash + 1));
        }
        return FilenameUtils.removeExtension(path);
    }

    public synchronized long getSize(String path) throws FileSystemException {
        path = sanitizePath(path);
        MountWrapper mount = getMount(path);
        return mount.getSize(path);
    }

    public synchronized String[] list(String path) throws FileSystemException {
        path = sanitizePath(path);
        MountWrapper mount = getMount(path);

        List<String> list = Lists.newArrayList();
        mount.list(path, list);

        for(MountWrapper otherMount : this.mounts.values()){
            if(getDirectory(otherMount.getLocation()).equals(path)){
                list.add(getName(otherMount.getLocation()));
            }
        }

        String[] array = new String[list.size()];
        list.toArray(array);
        return array;
    }

    private void findIn(String dir, List<String> matches, Pattern wildPattern) throws FileSystemException {
        String[] list = list(dir);
        //noinspection ForLoopReplaceableByForEach
        for(int i = 0; i < list.length; i++){
            String entry = list[i];
            String entryPath = dir + "/" + entry;
            if(wildPattern.matcher(entryPath).matches()){
                matches.add(entryPath);
            }
            if(isDir(entryPath)){
                findIn(entryPath, matches, wildPattern);
            }
        }
    }

    public synchronized String[] find(String wildPath) throws FileSystemException {
        wildPath = sanitizePath(wildPath, true);
        Pattern wildPattern = Pattern.compile("^\\Q" + wildPath.replaceAll("\\*", "\\\\E[^\\\\/]*\\\\Q") + "\\E$");
        List<String> matches = Lists.newArrayList();
        findIn("", matches, wildPattern);
        return matches.toArray(new String[matches.size()]);
    }

    public synchronized boolean exists(String path) throws FileSystemException {
        path = sanitizePath(path);
        MountWrapper mount = getMount(path);
        return mount.exists(path);
    }

    public synchronized boolean isDir(String path) throws FileSystemException {
        path = sanitizePath(path);
        MountWrapper mount = getMount(path);
        return mount.isDirectory(path);
    }

    public synchronized IMountedFileText openForRead(String path) throws FileSystemException {
        path = sanitizePath(path);
        MountWrapper mount = getMount(path);
        InputStream stream = mount.openForRead(path);
        if(stream != null){
            final BufferedReader reader = new BufferedReader(new InputStreamReader(stream));
            IMountedFileText file = new IMountedFileText() {
                public String readLine() throws IOException {
                    return reader.readLine();
                }

                public void write(String s, int off, int len, boolean newLine) throws IOException {
                    throw new UnsupportedOperationException();
                }

                public void close() throws IOException {
                    //noinspection SynchronizeOnNonFinalField
                    synchronized(FileSystem.this.openFiles){
                        FileSystem.this.openFiles.remove(this);
                        reader.close();
                    }
                }

                public void flush() throws IOException {
                    throw new UnsupportedOperationException();
                }
            };
            //noinspection SynchronizeOnNonFinalField
            synchronized(this.openFiles){
                this.openFiles.add(file);
            }
            return file;
        }
        return null;
    }

    public synchronized IMountedFileBinary openForBinaryRead(String path) throws FileSystemException {
        path = sanitizePath(path);
        MountWrapper mount = getMount(path);
        final InputStream stream = mount.openForRead(path);
        if(stream != null){
            IMountedFileBinary file = new IMountedFileBinary() {
                public int read() throws IOException {
                    return stream.read();
                }

                public void write(int i) throws IOException {
                    throw new UnsupportedOperationException();
                }

                public void close() throws IOException {
                    //noinspection SynchronizeOnNonFinalField
                    synchronized(FileSystem.this.openFiles){
                        FileSystem.this.openFiles.remove(this);
                        stream.close();
                    }
                }

                public void flush()
                        throws IOException {
                    throw new UnsupportedOperationException();
                }
            };
            //noinspection SynchronizeOnNonFinalField
            synchronized(this.openFiles){
                this.openFiles.add(file);
            }
            return file;
        }
        return null;
    }

    private MountWrapper getMount(String path)
            throws FileSystemException {
        Iterator it = this.mounts.values().iterator();
        MountWrapper match = null;
        int matchLength = 999;
        while(it.hasNext()){
            MountWrapper mount = (MountWrapper) it.next();
            if(contains(mount.getLocation(), path)){
                int len = toLocal(path, mount.getLocation()).length();
                if((match == null) || (len < matchLength)){
                    match = mount;
                    matchLength = len;
                }
            }
        }
        if(match == null){
            throw new FileSystemException("Invalid Path");
        }
        return match;
    }

    private static String sanitizePath(String path) {
        return sanitizePath(path, false);
    }

    private static String sanitizePath(String path, boolean allowWildcards) {
        path = path.replace('\\', '/');

        char[] specialChars = {'"', ':', '<', '>', '?', '|'};

        StringBuilder cleanName = new StringBuilder();
        for(int i = 0; i < path.length(); i++){
            char c = path.charAt(i);
            if((c >= ' ') && (Arrays.binarySearch(specialChars, c) < 0) && (allowWildcards || c != '*')){
                cleanName.append(c);
            }
        }
        path = cleanName.toString();

        String[] parts = path.split("/");
        Stack<String> outputParts = new Stack<String>();
        //noinspection ForLoopReplaceableByForEach
        for(int n = 0; n < parts.length; n++){
            String part = parts[n];
            if((part.length() != 0) && (!".".equals(part))){
                if("..".equals(part)){
                    if(!outputParts.empty()){
                        String top = outputParts.peek();
                        if(!"..".equals(top)){
                            outputParts.pop();
                        }else{
                            outputParts.push("..");
                        }
                    }else{
                        outputParts.push("..");
                    }
                }else if(part.length() >= 255){
                    outputParts.push(part.substring(0, 255));
                }else{
                    outputParts.push(part);
                }
            }
        }

        StringBuilder result = new StringBuilder("");
        Iterator it = outputParts.iterator();
        while(it.hasNext()){
            String part = (String) it.next();
            result.append(part);
            if(it.hasNext()){
                result.append('/');
            }
        }

        return result.toString();
    }

    public static boolean contains(String pathA, String pathB) {
        pathA = sanitizePath(pathA);
        pathB = sanitizePath(pathB);

        return !"..".equals(pathB) && !pathB.startsWith("../") && (pathB.equals(pathA) || pathA.length() == 0 || pathB.startsWith(pathA + "/"));
    }

    public static String toLocal(String path, String location) {
        path = sanitizePath(path);
        location = sanitizePath(location);

        assert contains(location, path);
        String local = path.substring(location.length());
        if(local.startsWith("/")){
            return local.substring(1);
        }
        return local;
    }

    private class MountWrapper {

        private String location;
        private IMount mount;

        public MountWrapper(String location, IMount mount) {
            Checks.notNull(location, "Location is null");
            Checks.notNull(mount, "Mount is null");

            this.location = location;
            this.mount = mount;
        }

        public boolean exists(String path) throws FileSystemException {
            path = toLocal(path);
            try{
                return this.mount.exists(path);
            }catch(IOException e){
                throw new FileSystemException(e.getMessage());
            }
        }

        public boolean isDirectory(String path) throws FileSystemException {
            path = toLocal(path);
            try{
                return (this.mount.exists(path)) && (this.mount.isDirectory(path));
            }catch(IOException e){
                throw new FileSystemException(e.getMessage());
            }
        }

        public void list(String path, List<String> contents) throws FileSystemException {
            path = toLocal(path);
            try{
                if(this.mount.exists(path) && this.mount.isDirectory(path)){
                    this.mount.list(path, contents);
                }else{
                    throw new FileSystemException("Not a directory");
                }
            }catch(IOException e){
                throw new FileSystemException(e.getMessage());
            }
        }

        public long getSize(String path) throws FileSystemException {
            path = toLocal(path);
            try{
                if(this.mount.exists(path)){
                    if(this.mount.isDirectory(path)){
                        return 0L;
                    }

                    return this.mount.getSize(path);
                }

                throw new FileSystemException("No such file");
            }catch(IOException e){
                throw new FileSystemException(e.getMessage());
            }
        }

        public InputStream openForRead(String path) throws FileSystemException {
            path = toLocal(path);
            try{
                if((this.mount.exists(path)) && (!this.mount.isDirectory(path))){
                    return this.mount.openForRead(path);
                }

                throw new FileSystemException("No such file");
            }catch(IOException e){
                throw new FileSystemException(e.getMessage());
            }
        }

        private String toLocal(String path) {
            return FileSystem.toLocal(path, this.location);
        }

        public String getLocation() {
            return this.location;
        }
    }
}
